File: src/main/java/com/szokone/aem/gokotlin/java/servlet/SimpleServlet.java
package com.szokone.aem.gokotlin.java.servlet;

import com.google.common.collect.ImmutableMap;
import com.google.common.net.MediaType;
import com.szokone.aem.gokotlin.java.service.SimpleService;
import org.apache.sling.api.SlingHttpServletRequest;
import org.apache.sling.api.SlingHttpServletResponse;
import org.apache.sling.api.servlets.SlingAllMethodsServlet;
import org.osgi.service.component.annotations.Component;
import org.osgi.service.component.annotations.Reference;

import javax.servlet.Servlet;
import java.io.IOException;

/**
 * OSGi component registered as Sling Servlet
 *
 * @link http://[host]:[port]/bin/java/hello
 * <p>
 * Other Sling Servlet properties:
 * <p>
 * sling.servlet.extensions=
 * sling.servlet.selectors=
 * sling.servlet.resourceTypes=
 * sling.servlet.methods=
 * sling.servlet.prefix=
 * <p>
 * More: https://sling.apache.org/documentation/the-sling-engine/servlets.html
 */
@Component(
        service = Servlet.class,
        property = {
                "sling.servlet.paths=/bin/java/hello"
        }
)
/*
  When using `org.apache.felix.scr.annotations` Felix SCR equivalent:

  @SlingServlet( paths = "/bin/java/hello" )
 */
public class SimpleServlet extends SlingAllMethodsServlet {

    @Reference
    private SimpleService simpleService;

    @Override
    protected void doGet(SlingHttpServletRequest request, SlingHttpServletResponse response) throws IOException {

        response.setContentType(MediaType.JSON_UTF_8.toString());
        response.getWriter().write(simpleService.toJson(ImmutableMap.of("message", "Hello from Java World!")));
    }

}

File: src/main/kotlin/com/szokone/aem/gokotlin/kotlin/servlet/SimpleServlet.kt
package com.szokone.aem.gokotlin.kotlin.servlet

import com.google.common.collect.ImmutableMap
import com.google.common.net.MediaType
import com.szokone.aem.gokotlin.kotlin.service.SimpleService
import org.apache.sling.api.SlingHttpServletRequest
import org.apache.sling.api.SlingHttpServletResponse
import org.apache.sling.api.servlets.SlingAllMethodsServlet
import org.osgi.service.component.annotations.Component
import org.osgi.service.component.annotations.Reference
import javax.servlet.Servlet

/**
 * @link http://host:port/bin/kotlin/hello
 */
@Component(
  service = arrayOf(Servlet::class),
  property = arrayOf("sling.servlet.paths=/bin/kotlin/hello")
)
class SimpleServlet : SlingAllMethodsServlet() {

  @Reference
  lateinit var simpleService: SimpleService

  override fun doGet(request: SlingHttpServletRequest, response: SlingHttpServletResponse) {

    response.contentType = MediaType.JSON_UTF_8.toString()
    response.writer.write(simpleService.toJson(ImmutableMap.of("message", "Hello from Kotlin World!")))
  }
}
